src/Button/Button.tsx

```diff
@@ -0,0 +1,227 @@
+import * as React from 'react';
+import cx from '../utils/classnames';
+import every = require('lodash/every');
+import { props, t, stateClassUtil } from '../utils';
+import _TextOverflow from '../TextOverflow/TextOverflow';
+import { PopoverProps } from '../Popover/Popover';
+import { ObjectOverwrite } from 'typelevel-ts';
+import FlexView from 'react-flexview';
+import Icon from '../Icon/Icon';
+import LoadingSpinner from '../LoadingSpinner/LoadingSpinner';
+
+export namespace ButtonProps {
+
+  export type ButtonState = 'ready' | 'not-allowed' | 'processing' | 'error' | 'success';
+  export type ButtonType = 'default' | 'primary' | 'positive' | 'negative' | 'flat';
+  export type ButtonSize = 'tiny' | 'small' | 'medium';
+
+}
+
+// TODO: TextOverflow should accept a popoverProps
+export type TextOverflowCompatibleComponent = React.ComponentClass<{
+  label?: string | number,
+  popover?: ObjectOverwrite<PopoverProps.Popover, {
+    position?: PopoverProps.Position,
+    content?: void & string
+  }>
+}>
+
+export interface ButtonProps {
+  /** callback */
+  onClick: (e: React.SyntheticEvent<HTMLDivElement>) => void,
+  /** ready, not-allowed, processing, success, error; overrides `baseState`, use it if you want button to be a functional component */
+  buttonState?: ButtonProps.ButtonState,
+  /** can be a String, or a dictionary { [buttonState]: String }, t.maybe(t.union([t.Str,  stringForButtonStates]) */
+  label?: string | { [key in ButtonProps.ButtonState]?: string },
+  /** can be a String referring to an icon, or a dictionary { [buttonState]: String },t.maybe(t.union([t.Str, stringForButtonStates])) */
+  icon?: string | { [key in ButtonProps.ButtonState]?: string },
+  /** otherwise just pass a string as children */
+  children?: string,
+  /** type of the button (default, primary, positive, negative, flat) */
+  type?: ButtonProps.ButtonType,
+  /** shortcut for type "primary" */
+  primary?: boolean,
+  /** shortcut for type "flat" */
+  flat?: boolean,
+  /** size of the button, one of 'tiny', 'small', 'medium' */
+  size?: ButtonProps.ButtonSize,
+  /** fluid - fluid (block) button, takes the width of the container */
+  fluid?: boolean,
+  /** circular - circular button, this is allowed only if it's an icon button */
+  circular?: boolean,
+  /** textOverflow - function to handle the overflow of too long labels, replacing with ellipsed string and tooltip */
+  textOverflow?: TextOverflowCompatibleComponent,
+  style?: React.CSSProperties,
```

Comment:
you can copy the description of `style`/`className` from the same props in `AsyncStatusIndicator`

Reply:
The description from `AsyncStatusIndicator` is not applicable to `Button` component as it says:
> /** an optional style object to pass to top level element of the component */

and, as I pointed out in the PR description, the `style` is assigned to the `.button-inner` element.
This is not a big deal, but it's incorrect anyway.

I'll rephrase as:
> /** an optional style object to pass to first inner element of the component */

is it better?

Reply:
sure!